google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,422 @@
+class CompletedOperationMetric:
+    """
+    A dataclass representing the data associated with a completed rpc operation.
+    """
+
+    op_type: OperationType
+    start_time: float
+    duration: float
+    op_id: UUID
+    completed_attempts: list[CompletedAttemptMetric]
+    final_status: StatusCode
+    cluster_id: str | None
+    zone: str
+    is_streaming: bool
+
+
+@dataclass
+class ActiveAttemptMetric:
+    start_time: float = field(default_factory=time.monotonic)
+    # the time it takes to recieve the first response from the server
+    # currently only tracked for ReadRows
+    first_response_latency: float | None = None
+    # the time taken by the backend. Taken from response header
+    gfe_latency: float | None = None
+
+
+@dataclass
+class ActiveOperationMetric:
+    """
+    A dataclass representing the data associated with an rpc operation that is
+    currently in progress.
+    """
+
+    op_type: OperationType
+    start_time: float = field(default_factory=time.monotonic)
+    op_id: UUID = field(default_factory=uuid4)
+    active_attempt: ActiveAttemptMetric | None = None
+    cluster_id: str | None = None
+    zone: str | None = None
+    completed_attempts: list[CompletedAttemptMetric] = field(default_factory=list)
+    handlers: list[MetricsHandler] = field(default_factory=list)
+    is_streaming: bool = False  # only True for read_rows operations
+    was_completed: bool = False
+
+    @property
+    def state(self) -> OperationState:
+        if self.was_completed:
+            return OperationState.COMPLETED
+        elif self.active_attempt is None:
+            if self.completed_attempts:
+                return OperationState.BETWEEN_ATTEMPTS
+            else:
+                return OperationState.CREATED
+        else:
+            return OperationState.ACTIVE_ATTEMPT
+
+    def start(self) -> None:
+        """
+        Optionally called to mark the start of the operation. If not called,
+        the operation will be started at initialization.
+
+        If the operation was completed or has active attempts, will raise an
+        exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+        """
+        if self.state != OperationState.CREATED:
+            return self._handle_error(INVALID_STATE_ERROR.format("start", self.state))
+        self.start_time = time.monotonic()
+
+    def start_attempt(self) -> None:
+        """
+        Called to initiate a new attempt for the operation.
+
+        If the operation was completed or there is already an active attempt,
+        will raise an exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+        """
+        if self.state != OperationState.BETWEEN_ATTEMPTS and self.state != OperationState.CREATED:
+            return self._handle_error(INVALID_STATE_ERROR.format("start_attempt", self.state))
+
+        self.active_attempt = ActiveAttemptMetric()
+
+    def add_call_metadata(self, metadata: dict[str, bytes | str]) -> None:
```

Comment:
please name this `add_response_metadata` to avoid confusion with request headers